rewrite-maven/src/main/java/org/openrewrite/maven/MavenParser.java

```diff
@@ -81,10 +81,13 @@ public Stream<SourceFile> parseInputs(Iterable<Input> sources, @Nullable Path re
                 if (pom.getProperties().isEmpty()) {
                     pom = pom.withProperties(new LinkedHashMap<>());
                 }
+
                 String baseDir = pomPath.toAbsolutePath().getParent().toString();
-                pom.getProperties().put("project.basedir", baseDir);
-                pom.getProperties().put("basedir", baseDir);
-                pom.getProperties().putAll(properties);
+                pom = pom.withConstants(new HashMap<String, String>() {{
+                    put("project.basedir", baseDir);
+                    put("basedir", baseDir);
+                    putAll(properties);
```

Comment:
Should the properties really be considered as constants? They can change from recipe to recipe.

Reply:
Uhm I am not absolutely sure, but these 'external properties' should never be changed by a recipe, right?

Reply:
Perhaps good to follow the Maven naming here; these are user properties coming from the MavenSession, and could be coming from .mvn/maven.config, or explicit arguments passed in. That's also why I'd want to get them from the session in the plugin, as opposed to from parsing the config file (same with profiles: use the session in the plugin instead there too).
https://maven.apache.org/ref/3.9.9/maven-core/apidocs/org/apache/maven/execution/MavenSession.html#getUserProperties()